### PR TITLE
Feature fix websockets

### PR DESCRIPTION
**Pull request by T-mclennan**
Greetings! The goal of this feature branch was to add functionality and security to Socket.io logic, as per issue #3.

**New Features:**
- I added functionality to give a notifications when a user begins and stops typing. 
- Typing notification is used to trigger animation inside live chat, and an italicized 'typing...' notification on sidebar. 
- I added a listener for disconnect, that lets others know that the user has gone offline when window is closed. 

**Security Improvements**
- I added logic to create a unique UUID on the frontend as a 'socketId', and persist it in local storage. 
- The socketId is used to make a unique room for every user, allowing the server to send out targeted messages and notifications through socket.io to one specific recipient. 
- On the backend the socketId is stored inside a Map of online users, with the userId as the key. 
- Incoming requests that don't want a general audience (like messages and notifications for a conversation) are sent to specific recipients, while general notifications are broadcast publicly. 
- An authentication layer is added with the help of the socketio-auth package, which validates the user token with JWT in a similar manner to user authentication.

### CODE REVIEW

File: client/src/actions/thunkCreators.js [hunk 1]
```diff
@@ -94,15 +94,29 @@ const sendMessage = (data, body) => {
...
     sendMessage(data, body);
+    broadcastTypingAction(conversationId, recipientId, 'stoppedTyping');
+  } catch (error) {
+    console.error(error);
+  }
+};
+
+export const markConvoAsSeen = (id) => async (dispatch) => {
+  try {
+    const { data } = await axios.post(`/api/conversations/viewed/${id}`);
```
[the remark this answers is not in the document]
Comment (T-mclennan) [reply client/src/actions/thunkCreators.js:114]: Absolutely, thanks Sebin.

File: server/db/models/conversation.js [hunk 2]
```diff
@@ -30,9 +34,40 @@ Conversation.createConversation = async function (senderId, recipientId) {
   conversation = await Conversation.create({
     user1Id: senderId,
```
[the remark this answers is not in the document]
Comment (T-mclennan) [reply server/db/models/conversation.js:35]: Yes we absolutely can track the messages if you think it's an improvement! I think I'm missing something in my understanding though, how can we set that up so it's efficient? This is how I approached it initially, tracking a flag in each message. It seemed like we'd have to touch every item in the Messages list to count the amount of unseen messages, and do that with almost every user interaction. I figured that we really only needed the total number for the whole conversation, and just incrementing and clearing that value is vastly less computation than having to repeatedly update and count the state of each message. I think I'm understanding this the wrong way, do you mind elaborating a bit more on what you're thinking?

File: client/src/socket.js [hunk 3]
```diff
@@ -1,26 +1,72 @@
 import io from "socket.io-client";
 import store from "./store";
+import { v4 as uuidv4 } from 'uuid';
```
[the remark this answers is not in the document]
Comment (T-mclennan) [reply client/src/socket.js:3]: Great idea, I'll change that!

File: server/bin/www [hunk 4]
```diff
@@ -28,32 +29,73 @@ const server = http.createServer(app);
...
+//Authentication middleware for socket.io
+require('socketio-auth')(io, {
+  authenticate: validateSocket
+});
+
 io.on("connection", (socket) => {
+
+  //Create a private room for each user denoted by socketId. 
+  const {socketId, token} = socket.handshake.query;
+  socket.join(socketId);
```
Comment (0916dhkim) [inline server/bin/www:44]: There are a couple concerns with this implementation:
1. Room name is passed-in from the client side so a malicious client can pass-in other user's room name to receive unauthorized socket events.
2. Room name is stored inside local storage and removed on logout. That's an additional sensitive information to manage on client side.

I suggest using user ID as a room name here.

File: server/onlineUsers.js [hunk 5]
```diff
@@ -1,2 +1,2 @@
-const onlineUsers = [];
+const onlineUsers = new Map();
```
Comment (0916dhkim) [inline server/onlineUsers.js:1]: Great performance improvement on online user tracking :100:

File: server/bin/www [hunk 6]
```diff
@@ -28,32 +29,73 @@ const server = http.createServer(app);
...
-      message: data.message,
-      sender: data.sender,
-    });
+    const {message, recipientId, sender } = data;
+    if (recipientId) {
+      const destinationId = onlineUsers.get(recipientId)
+      socket.broadcast.to(destinationId).emit("new-message", {
+        message,
+        sender,
+      });
```
Comment (0916dhkim) [inline server/bin/www:63]: Nice job :+1:

**Review by 0916dhkim — APPROVED** [review]
@T-mclennan Good job addressing Sebin's comments :+1: 
I have some concern about the implementation of socket room name generation. This PR looks great otherwise, so please merge them after addressing my comment below :point_down:

File: server/bin/www (hunk 4 above)
Comment (T-mclennan) [reply server/bin/www]: Those are really good points. Thank you!

**Pull request merged**